brave/src/main/java/brave/propagation/B3SingleFormat.java

```diff
@@ -0,0 +1,168 @@
+package brave.propagation;
+
+import brave.internal.Nullable;
+import java.util.Collections;
+import java.util.logging.Logger;
+
+import static brave.internal.HexCodec.lenientLowerHexToUnsignedLong;
+import static brave.internal.HexCodec.writeHexLong;
+import static brave.internal.TraceContexts.FLAG_DEBUG;
+import static brave.internal.TraceContexts.FLAG_SAMPLED;
+import static brave.internal.TraceContexts.FLAG_SAMPLED_SET;
+import static java.util.logging.Level.FINE;
+
+/** Implements the progation format described in {@link B3SinglePropagation}. */
+final class B3SingleFormat {
+  static final Logger logger = Logger.getLogger(B3SingleFormat.class.getName());
+  static final int FORMAT_MAX_LENGTH = 32 + 1 + 16 + 2 + 16 + 2; // traceid128-spanid-1-parentid-1
+
+  static String writeB3SingleFormat(TraceContext context) {
+    char[] result = getCharBuffer();
+    int pos = 0;
+    long traceIdHigh = context.traceIdHigh();
+    if (traceIdHigh != 0L) {
+      writeHexLong(result, pos, traceIdHigh);
+      pos += 16;
+    }
+    writeHexLong(result, pos, context.traceId());
+    pos += 16;
+    result[pos++] = '-';
+    writeHexLong(result, pos, context.spanId());
+    pos += 16;
+
+    Boolean sampled = context.sampled();
+    if (sampled != null) {
+      result[pos++] = '-';
+      result[pos++] = (sampled == Boolean.TRUE) ? '1' : '0';
+    }
+
+    long b3Id = context.parentIdAsLong();
+    if (b3Id != 0) {
+      result[pos++] = '-';
+      writeHexLong(result, pos, b3Id);
+      pos += 16;
+    }
+
+    if (context.debug()) {
+      result[pos++] = '-';
+      result[pos++] = '1';
+    }
+    return new String(result, 0, pos);
+  }
+
+  static @Nullable TraceContextOrSamplingFlags maybeB3SingleFormat(String b3) {
+    int length = b3.length();
+    if (length == 1) {
+      int flags = parseSampledFlag(b3, 0);
+      if (flags == 0) return null;
+      return TraceContextOrSamplingFlags.create(SamplingFlags.toSamplingFlags(flags));
+    } else if (length == 3) {
+      int flags = parseSampledFlag(b3, 0);
+      if (flags == 0) return null;
+      flags = parseDebugFlag(b3, 2, flags);
+      if (flags == 0) return null;
+      return TraceContextOrSamplingFlags.create(SamplingFlags.toSamplingFlags(flags));
+    }
+
+    // At this point we minimally expect a traceId-spanId pair
+    if (length < 16 + 1 + 16 /* traceid64-spanid */) {
+      logger.fine("Invalid input: truncated");
+      return null;
+    } else if (length > FORMAT_MAX_LENGTH) {
+      logger.fine("Invalid input: too long");
+      return null;
+    }
+
+    int pos = 0;
+    long traceIdHigh, traceId;
+    if (b3.charAt(32) == '-') {
+      traceIdHigh = lenientLowerHexToUnsignedLong(b3, 0, 16);
+      traceId = lenientLowerHexToUnsignedLong(b3, 16, 32);
```

Comment:
You do `lenientLowerHexToUnsignedLong(b3, pos, pos + 16);` for parsing the span ID, did you mean to use `pos` here also to be consistent?

Reply:
I probably meant to do the following in order to save from doing extra math when there will be constant results (due to this being the initial parse). Maybe the distraction isn't worth it.. I can switch to `pos, pos + 16` NP

```java
int pos;
if (b3.charAt(32) == '-') {
  --snip--
  pos = 33; // traceid128-
} else {
  --snip--
  pos = 17; // traceid64-
}
```

Reply:
I moved initialization of pos further down to make it ideally more clear